Add Schism eval benchmark proposal

wingo opened the pull request:
This is a draft PR to test the benchmark submission process; see issues #1 and #2.  What do you think, Ben?  Feedback of any sort very welcome.

wingo commented:
/cc @eholk

titzer commented on proposals/schism-eval-fib/README.md:
```diff
@@ -0,0 +1,204 @@
+ * We could use GC types to avoid calling out so much to the host.
+   Would need implementation of GC in an engine though.
+
+## Benchmark kind
```
What about "category" instead of kind?

wingo replied:
Sure.  FWIW I imagine at some point that there will be a centralized list of benchmarks and that a PR would add the benchmark to the right category on the centralized list.

titzer commented on proposals/schism-eval-fib/README.md:
```diff
@@ -0,0 +1,204 @@
+
+## Benchmark kind
+
+I think we can consider this benchmark to be a "kernel".
```
How big is the module that is generated? Am I right in assuming that the Schism implementation compiles the fib program to a single module, and that that module doesn't include much of the language runtime?

titzer replied:
Ah, I see the binary is ~20KiB. That's probably kernel territory.

It'd be good for us to think over the definition of what a kernel is. E.g. does it have a particularly small "working set size" in terms of number of functions or bytecodes?

wingo replied:
Right.  Schism currently compiles by first adding the user code to a set of primitives, and does not do dead-code elimination.  This would matter for other benchmarks but given that `eval.ss` itself adds most of those primitives to the default environment, I don't think DCE would change much here.

I am not certain what percentage of functions will be executed for this particular input (the fibonacci function).  Probably a fair bit just given that `eval` needs to parse and expand the input.

I think the problem is to have rules to differentiate a kernel from a microbenchmark.  Probably before merging anything we should add guidelines about how to categorize a given benchmark.  Actually, what is the purpose of differentiating between kernels and microbenchmarks?

titzer commented on proposals/schism-eval-fib/README.md:
````diff
@@ -0,0 +1,204 @@
+following:
+
+```
+75025
````
Excellent. Thanks for checking outputs 👍

titzer commented on proposals/schism-eval-fib/README.md:
````diff
@@ -0,0 +1,204 @@
+On my system this test runs in about 4 or 5 seconds:
+
+```
+real	0m4.729s
````
Cool. I assume since this is fib that configuring different inputs for different benchmark sizes.

What do you think of adding recommending small, medium and large inputs?

wingo replied:
Right, no problem; will do tomorrow.

Do you want benchmarks in a machine-readable format somehow?

Would you like a benchmark runner script, or do you have one in mind?

wingo commented:
I updated the branch to add a benchmark runner; see `runner.js` in `proposals/`.  I also updated the runner to not require Node, and to work with SpiderMonkey at least (still updating my JSC here).

Also, this benchmark now uses the tail call proposal in addition to the reference types proposal.
